delta-app/src/main/java/io/cdap/delta/app/CapacityBoundedEventQueue.java

```diff
@@ -0,0 +1,179 @@
+/*
+ *
+ * Copyright © 2021 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.cdap.delta.app;
+
+import io.cdap.delta.api.ChangeEvent;
+import io.cdap.delta.api.DDLEvent;
+import io.cdap.delta.api.DMLEvent;
+import io.cdap.delta.api.Sequenced;
+
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+
+  /**
+   * A capacity bounded event queue. It will block if the capacity of the events in the queue reaches certain
+   * limit. When calculating the capacity of an event only the size of the row and previous row in DML event are
+   * calculated. Because other fields are fixed size and could be limited by queue size (the number of event in the
+   * queue)
+   */
+  public class CapacityBoundedEventQueue extends ArrayBlockingQueue<Sequenced<? extends ChangeEvent>> {
+
+    private final long capacityLimit;
+    private long capacity;
+    /** Main lock guarding all access */
+    private final ReentrantLock lock;
+    /** Condition for waiting puts */
+    private final Condition notFull;
+
+    public CapacityBoundedEventQueue(int queueSize, long capcityLimit) {
+      super(queueSize);
+      this.capacityLimit = capcityLimit;
+      this.capacity = 0;
+      this.lock = new ReentrantLock();
+      this.notFull = lock.newCondition();
+    }
+
+    private static void checkNotNull(Sequenced<? extends ChangeEvent> event) {
+      if (event == null) {
+        throw new NullPointerException();
+      }
+    }
+
+    private static int computeCapacity(Sequenced<? extends ChangeEvent> event) {
+      if (event == null) {
+        return 0;
+      }
+      ChangeEvent changeEvent = event.getEvent();
+      if (changeEvent instanceof DDLEvent) {
+        return 0;
+      }
+      return ((DMLEvent) changeEvent).getSizeInBytes();
+    }
+
+    @Override
+    public void put(Sequenced<? extends ChangeEvent> event) throws InterruptedException {
+      checkNotNull(event);
+      final ReentrantLock lock = this.lock;
+      lock.lockInterruptibly();
+      try {
+        int eventCapacity = computeCapacity(event);
+        while (capacity + eventCapacity > capacityLimit) {
```

Comment:
if a single event is larger than the capacity limit, this will wait forever. Seems like it would be better to always try when the queue is empty. 

Immediately throwing an exception would also be appropriate if we just consider this class, but it wouldn't be nice for the replicator, since it would get into a continuous retry loop.

Reply:
good point :)  done :)